### PR TITLE
refactor: add automatic conversion for `nix::Errno`

### LINKED ISSUE #4007: Add automatic conversion from `nix::Errno` to `UError` like we have for `std::io::Error`.
We should probably add automatic conversion from `nix::Errno` to `UError` like we have for `std::io::Error`. Then we don't need these manual conversions.

https://docs.rs/uucore/latest/uucore/error/trait.UError.html

_Originally posted by @tertsdiepraam in https://github.com/uutils/coreutils/pull/4003#discussion_r986575861_

### PR DESCRIPTION
**Pull request by orhun**
An attempt to fix #4007

### CODE REVIEW

**Comment by sylvestre** [comment]
it would be nice to test this :)
thanks

File: src/uucore/src/lib/mods/error.rs [hunk 1]
```diff
@@ -508,6 +508,17 @@ impl From<std::io::Error> for Box<dyn UError> {
     }
 }
 
+impl<T> FromIo<UResult<T>> for Result<T, nix::Error> {
```
Comment (sylvestre) [inline src/uucore/src/lib/mods/error.rs:511]: needs some comments here :)
and usage examples

Comment (sylvestre) [reply src/uucore/src/lib/mods/error.rs:511]: see the function below

**Comment by tertsdiepraam** [comment]
Cool! This is exactly what I had in mind. I'd just like to have it implemented for `nix::error` (without being wrapped in a result) as well. (And I agree with Sylvestre's comments)

**Comment by orhun** [comment]
> it would be nice to test this :)

How can I test it? Is adding usage examples as comments enough?

> I'd just like to have it implemented for `nix::error` (without being wrapped in a result) as well.

Do you mean `impl<T> FromIo<UResult<T>> for nix::Error` or `impl FromIo<UIoError> for nix::Error`? Both are possible and I don't know which one is more convenient.

**Comment by tertsdiepraam** [comment]
> How can I test it? Is adding usage examples as comments enough?

You could make a unit test that you know gives an error and assert that the error that you get is correct.
 
> Do you mean `impl<T> FromIo<UResult<T>> for nix::Error` or `impl FromIo<UIoError> for nix::Error`? Both are possible and I don't know which one is more convenient.

The same as for `std::io::Error`. We map to `Box<UIoError>` because that's what goes into a `UResult`:
```rust
impl FromIo<Box<UIoError>> for std::io::Error {
    fn map_err_context(self, context: impl FnOnce() -> String) -> Box<UIoError> {
        Box::new(UIoError {
            context: Some((context)()),
            inner: self,
        })
    }
}
impl<T> FromIo<UResult<T>> for std::io::Result<T> {
    fn map_err_context(self, context: impl FnOnce() -> String) -> UResult<T> {
        self.map_err(|e| e.map_err_context(context) as Box<dyn UError>)
    }
}
```

These would also be nice to have for `nix::error`:
```rust
impl From<std::io::Error> for UIoError {
    fn from(f: std::io::Error) -> Self {
        Self {
            context: None,
            inner: f,
        }
    }
}
impl From<std::io::Error> for Box<dyn UError> {
    fn from(f: std::io::Error) -> Self {
        let u_error: UIoError = f.into();
        Box::new(u_error) as Self
    }
}
```

**Comment by sylvestre** [comment]
Could you please fix the clippy warning? thanks

File: src/uucore/src/lib/mods/error.rs (hunk 1 above)
Comment (orhun) [reply src/uucore/src/lib/mods/error.rs:526]: [c19c19e](https://github.com/uutils/coreutils/pull/4060/commits/c19c19e4db56678035d4aca4053128dfbcae7ed4)

**Comment by orhun** [comment]
I'm not sure why the CI is failing.

**Comment by orhun** [comment]
Friendly ping 🏓

**Comment by sylvestre** [comment]
Just rebased it. It had some job failures

**Comment by sylvestre** [comment]
Could you please have a look to https://github.com/uutils/coreutils/actions/runs/3355920132/jobs/5560563514?

**Comment by sylvestre** [comment]
ping? :)

**Comment by orhun** [comment]
Still not sure why the CI is failing 🤔

**Comment by sylvestre** [comment]
you can ignore these :)

**Pull request merged**